nvflare/apis/utils/reliable_message.py

```diff
@@ -0,0 +1,369 @@
+                self.executor.submit(self._do_request, request, fl_ctx)
+                return _status_reply(STATUS_IN_PROCESS)  # ack
+            elif self.result:
+                # we already finished processing - send the result back
+                return self.result
+            else:
+                # we are still processing
+                return _status_reply(STATUS_IN_PROCESS)
+        elif op == OP_QUERY:
+            if self.result:
+                if self.reply_time:
+                    # result already sent back successfully
+                    return _status_reply(STATUS_REPLIED)
+                elif self.replying:
+                    # result is being sent
+                    return _status_reply(STATUS_IN_REPLY)
+                else:
+                    # try to send the result again
+                    return self.result
+            else:
+                # still in process
+                if time.time() - self.rcv_time > self.timeout:
+                    # the process is taking too much time
+                    return _status_reply(STATUS_ABORTED)
+                else:
+                    return _status_reply(STATUS_IN_PROCESS)
+
+    def _try_reply(self, fl_ctx: FLContext):
+        engine = fl_ctx.get_engine()
+        self.replying = True
+        ack = engine.send_aux_request(
+            targets=[self.source],
+            topic=TOPIC_RELIABLE_REPLY,
+            request=self.result,
+            timeout=self.timeout,
+            fl_ctx=fl_ctx,
+        )
+        self.replying = False
+        _, rc = _extract_result(ack, self.source)
+        if rc == ReturnCode.OK:
+            # reply sent successfully!
+            self.reply_time = time.time()
+
+    def _do_request(self, request: Shareable, fl_ctx: FLContext):
+        try:
+            result = self.request_handler_f(self.topic, request, fl_ctx)
+        except Exception as e:
+            result = _error_reply(ReturnCode.EXECUTION_EXCEPTION, str(e))
+
+        # send back
+        result.set_header(HEADER_TX, self.tx_id)
+        result.set_header(HEADER_OP, OP_REPLY)
+        result.set_header(HEADER_TOPIC, self.topic)
+        self.result = result
+        self._try_reply(fl_ctx)
+
+
+class _ReplyReceiver:
+    def __init__(self, tx_id: str):
+        self.tx_id = tx_id
+        self.result = None
+        self.result_ready = threading.Event()
+
+    def process(self, reply: Shareable) -> Shareable:
+        self.result = reply
+        self.result_ready.set()
+        return make_reply(ReturnCode.OK)
+
+
+class ReliableMessage:
+
+    _topic_to_handle = {}
+    _req_receivers = {}  # tx id => receiver
+    _enabled = False
+    _executor = None
+    _query_interval = 1.0
+    _max_retries = 5
+    _max_tx_time = 300.0  # 5 minutes
+    _reply_receivers = {}  # tx id => receiver
+    _tx_lock = threading.Lock()
+    _shutdown_asked = False
+
+    @classmethod
+    def register_request_handler(cls, topic: str, handler_f):
```

Comment:
might be good to add docstring about what "handler_f" needs to be

Reply:
Added DocString.